app/src/server/api/external/v1Api.router.ts

```diff
@@ -334,6 +347,135 @@ export const v1ApiRouter = createOpenApiRouter({
+          .describe(
+            'Extra tags to attach to the call for filtering. Eg { "userId": "123", "promptId": "populate-title" }',
+          ),
+      }),
+    )
+    .output(z.object({ matchedLogs: z.number() }))
+    .mutation(async ({ input, ctx }) => {
+      if (ctx.key.readOnly) {
+        throw new TRPCError({
+          message: "Read-only API keys cannot update log tags",
+          code: "FORBIDDEN",
+        });
+      }
+
+      let tags: Record<string, string | null> = {};
+      try {
+        tags = parseTags(input.tags, true);
+      } catch (e) {
+        throw new TRPCError({
+          message: `Failed to parse tags: ${(e as Error).message}`,
+          code: "BAD_REQUEST",
+        });
+      }
+
+      const tagNamesToDelete = Object.keys(tags).filter((key) => tags[key] === null);
+      const tagsToUpsert = Object.entries(tags).filter(([, value]) => value !== null) as [
+        string,
+        string,
+      ][];
+
+      const filters: z.infer<typeof filtersSchema> = [];
+
+      for (const filter of input.filters) {
+        filters.push({
+          field: filter.field,
+          comparator: "=",
+          value: filter.equals.toString(),
+        });
+      }
+
+      const matchedLogs = await constructLoggedCallFiltersQuery({
+        filters,
+        projectId: ctx.key.projectId,
+      })
+        .select(sql<number>`count(*)::int`.as("count"))
+        .executeTakeFirst();
+
+      if (tagNamesToDelete.length > 0) {
+        await kysely
+          .deleteFrom("LoggedCallTag")
+          .using((eb) =>
+            constructLoggedCallFiltersQuery({
+              filters,
+              projectId: ctx.key.projectId,
+              lctEB: eb,
+            })
+              .select("lc.id")
+              .as("lc"),
+          )
+          .whereRef("LoggedCallTag.loggedCallId", "=", "lc.id")
+          .where("LoggedCallTag.name", "in", tagNamesToDelete)
+          .execute();
+      }
+
+      const loggedCalls = await constructLoggedCallFiltersQuery({
+        filters,
+        projectId: ctx.key.projectId,
+      })
+        .select("lc.id")
+        .execute();
+
+      const dataToInsert: {
+        id: string;
+        name: string;
+        value: string;
+        projectId: string;
+        loggedCallId: string;
+      }[] = [];
+
+      // Iterate over each logged call and insert tags
+      for (const loggedCall of loggedCalls) {
+        const loggedCallId = loggedCall.id;
+
+        // Prepare the insert data for each tag
+        dataToInsert.push(
+          ...tagsToUpsert.map(([name, value]) => ({
+            id: uuidv4(),
+            name,
+            value,
+            projectId: ctx.key.projectId,
+            loggedCallId,
+          })),
+        );
+      }
+
+      if (dataToInsert.length) {
+        await kysely
+          .insertInto("LoggedCallTag")
+          .columns(["name", "value", "projectId", "loggedCallId"])
```

Comment:
This is actually just my ignorance about kysely since I assume this works, but why do you set `id: uuidv4()` above if it isn't one of the columns you're inserting into here?

Reply:
That is a valid question. Empirically, if I don't include an id for each tag, this operation fails. However, I'm not sure why this operation works while "id" is excluded from the enumerated columns into which we are inserting.